fetch-in-react-practice/client/components/app.jsx

```diff
@@ -0,0 +1,89 @@
+import React from 'react';
+import PageTitle from './page-title';
+import TodoList from './todo-list';
+import TodoForm from './todo-form';
+
+export default class App extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      todos: []
+    };
+    this.addTodo = this.addTodo.bind(this);
+    this.toggleCompleted = this.toggleCompleted.bind(this);
+  }
+
+  componentDidMount() {
+    fetch('/api/todos')
+      .then(response => response.json())
+      .then(data => {
+        const todoList = [];
+        for (let i = 0; i < data.length; i++) {
+          todoList.push(data[i]);
+        }
+        this.setState({ todos: todoList });
+      });
+
+  }
+
+  addTodo(newTodo) {
+    fetch('/api/todos', {
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/json'
+      },
+      body: JSON.stringify(newTodo)
+    })
+      .then(response => response.json())
+      .then(newTodo => {
+        const todoList = this.state.todos.concat(newTodo);
+        this.setState({ todos: todoList });
+      }
+      );
+
+  }
+
+  toggleCompleted(todoId) {
+    let todo;
+
+    for (let i = 0; i < this.state.todos.length; i++) {
+      if (this.state.todos[i].todoId === todoId) {
+        todo = i;
+      }
+
+    }
+
+    const opposite = { isCompleted: !this.state.todos[todo].isCompleted };
+
+    fetch(`/api/todos/${todoId}`, {
+      method: 'PATCH',
+      headers: {
+        'Content-Type': 'application/json'
+      },
+      body: JSON.stringify(opposite)
+    }
+
+    )
+      .then(response => response.json())
+      .then(data => {
+        const todoList3 = this.state.todos;
```

Comment:
This won't make a copy of the array, it just creates another reference to the same array. So on the next line when you make changes to the array you are mutating the state which can cause your app to not update correctly and/or corrupt your data in state. To fix this you need to make an actual copy of the array, below are 2 common ways to copy an array.

Using [slice](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/slice)
```JAVASCRIPT
const todoList3 = this.state.todos.slice();
```

Using ES6 [spread operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax)
```JAVASCRIPT
const todoList3 = [...this.state.todos];
```

Reply:
Thank you so much. Used the spread operator!